TODO-003 Change task status

jatanolu opened the pull request:
Acceptance Criteria

Given a list of tasks
When an individual task is displayed
Then a checkbox is displayed to the right of the task
And the check box should default to not checked

Given a task
When the checkbox is checked
The the task is complete

Given a task
When the checkbox is unchecked
The the task is pending

Given a list of tasks
When the tasks are displayed
Then the completed tasks are shown in a separate section below the pending task list

Given a task
When the status changes between complete and pending
Then the task moves to the appropriate section

jatanolu committed:
TODO-003 Change task status

Kotauror commented on backend/views.py:
```diff
@@ -21,3 +21,13 @@ def addTask(request):
+@api_view(['PUT'])
+def updateTask(request):
+    try:
+        updatedTask = Task.objects.get(id=request.data['id'])
```
more like taskToUpdate, it's not updated yet :)

jay-meister replied:
@Kotauror's comment was about the variable that you have defined on line 28, not the route handler name on line 26.

```py
taskToUpdate = Task.objects.get(id=request.data['id'])
```

Kotauror commented on frontend/components/Status.tsx:
```diff
@@ -0,0 +1,65 @@
+import { useEffect, useState } from "react";
+import { ITask, Ilist } from "../src/App";
+
+function Status(props: {
```
I think the name should be changed as this component does much more than rendering the `Status`.

Kotauror commented on frontend/components/Status.tsx:
```diff
@@ -0,0 +1,65 @@
+}) {
+  const [complete, setComplete] = useState<ITask[]>([]);
+  const [pending, setPending] = useState<ITask[]>([]);
+  function splitList() {
```
I'd rename this one - the name `splitList` describes the programming operation, or in other words, how you achieve the goal. Naming should reflect _what_ is your goal instead of _how_ you implemented it. In this case this could be something like `divideTasksAccordingToTheirCompletedState`. Feel free to come up with your own if mine seems too wordy.

jay-meister replied:
I could be wrong as I have not been able to get it running locally, but I feel like we can remove this function altogether. 

From what I can tell, we do not need to store `completedTasks` and `pendingTasks` in state, we just want to take the `taskList` and filter depending on the value of `completed`. I would do this directly in with the JSX in the return statement.

See my code snippet the comment further down in this file.

Kotauror commented on frontend/components/Status.tsx:
```diff
@@ -0,0 +1,65 @@
+  }
+  async function handleChange(id: number) {
+    let updateTask = await axios.put<Ilist>("updateTask/", { id: id });
+    console.log(updateTask);
```
remove console.log

Kotauror commented on frontend/components/Status.tsx:
```diff
@@ -0,0 +1,65 @@
+    splitList();
+  }, [props.list]);
+  return (
+    <>
+      <h2>Pending</h2>
+      {pending.map((task) => {
+        return (
+          <div className="task">
+            <p>{task.title}</p>
+            <input
+              className="checkbox"
+              type="checkbox"
+              onChange={() => handleChange(task.id)}
+              checked={task.completed}
+            />
+          </div>
+        );
+      })}
+      <h2>Completed</h2>
+      {complete.map((task) => {
+        return (
+          <div className="task">
+            <p>{task.title}</p>
+            <input
+              className="checkbox"
+              type="checkbox"
+              onChange={() => handleChange(task.id)}
+              checked={task.completed}
+            />
+          </div>
+        );
+      })}
+    </>
```
Lines 34-47 and 48-61 hardly differ - only by the headers and the list on #35 and #49 - please extract the logic into a new component and use it in both places.

jay-meister replied:
Agreed, although personally I would just extract the `TaskWithCheckbox` piece of the logic (which are identical in the above `map`s), and keep the `h2` in place.

That's just personal preference and totally up to you though.

eg:
```jsx
<h2>Pending</h2> 
{taskList.filter(task => !task.completed).map(task => <TaskWithCheckbox />)} 

<h2>Completed</h2> 
{taskList.filter(task => task.completed).map(task => <TaskWithCheckbox />)}
```

Kotauror reviewed (changes requested):
Blocked as there are no tests on the FE and the BE.

jay-meister commented on frontend/components/Task.tsx:
```diff
@@ -30,10 +28,7 @@ function Task(props: {
-      {props.list.map((task) => {
-        return <p>{task.title}</p>;
-      })}
+      <Status {...props} />
```
You shouldn't need to wrap this <Status /> in an empty `fragment` (`<> </>`)

jay-meister commented on frontend/components/Task.tsx:
```diff
@@ -1,19 +1,17 @@
-
   const tasklist = async () => {
     let allTasks = await axios.get<Ilist>("getTasks/");
+    console.log(allTasks);
```
🔦  Another log 🪵

jay-meister commented on frontend/components/Status.tsx:
```diff
@@ -0,0 +1,65 @@
+import axios from "axios";
+import { useEffect, useState } from "react";
+import { ITask, Ilist } from "../src/App";
```
These types are lacking consistency of capitalisation: `ITask` vs `Ilist` (should be `IList`).

More generally, I don't know what the `I` represents in `ITask` and `IList`, but it seems like it _could_ be redundant, so just wanted to flag that.

jay-meister commented on frontend/components/Status.tsx:
```diff
@@ -0,0 +1,65 @@
+function Status(props: {
+  submited: boolean;
+  setSubmited: (Submit: boolean) => void;
+  list: ITask[];
```
I also find the name `list` unclear. Many things in the codebase are 'lists'. A more explicit name would be `taskList` or simply `tasks`.

jay-meister commented on frontend/components/Status.tsx:
```diff
@@ -0,0 +1,65 @@
+  const [complete, setComplete] = useState<ITask[]>([]);
+  const [pending, setPending] = useState<ITask[]>([]);
+  function splitList() {
+    for (let x in props.list) {
```
In line with my previous naming comment, I'd expect to see more explicit naming here:

eg: `for (let task in props.taskList) {`

jay-meister commented on frontend/components/Status.tsx:
```diff
@@ -0,0 +1,65 @@
+    }
+  }
+  async function handleChange(id: number) {
+    let updateTask = await axios.put<Ilist>("updateTask/", { id: id });
```
Here we create a variable: `let updateTask` but do not use it.

I'd expect you to check the response from the backend, and if the status code is 200, then we can let the user know.

jay-meister commented on backend/views.py:
```diff
@@ -21,3 +21,13 @@ def addTask(request):
     return JsonResponse({'success': True})
+
+@api_view(['PUT'])
+def updateTask(request):
```
I don't know python or django, but I think this path and route handler should be called something more specific to what it actually does.

eg. the path would be something like: `/toggle-task-completed` which maps to the `toggleTaskCompleted` route handler

jay-meister commented on frontend/components/Status.tsx:
```diff
@@ -0,0 +1,65 @@
+  return (
+    <>
+      <h2>Pending</h2>
+      {pending.map((task) => {
```
Sorry to keep banging the same drum, but I feel like naming here could be more explicit: `pendingTasks` & `completedTasks`

jay-meister commented on frontend/components/Status.tsx:
```diff
@@ -0,0 +1,65 @@
+import { ITask, Ilist } from "../src/App";
+
+function Status(props: {
+  submited: boolean;
+  setSubmited: (Submit: boolean) => void;
```
slight spelling error/typo: `submitted`

jay-meister reviewed (changes requested):
Nice work @jatanolu 👏 

There are a lot of comments here, but most of mine are naming related and therefore quick changes. I also believe there's a bit of unnecessary work being done in the `Status` component.

I wasn't able to get it running so I couldn't validate my comments unfortunately.

Please message in slack if you have any questions. Once our review comments are addressed, the PR will be simplified and we can see if there are any further simplifications that can be made ✨ 


___
Further small point:

<img width="400" alt="Screenshot 2023-02-21 at 10 05 30" src="https://user-images.githubusercontent.com/12638261/220313964-60c24ff1-a891-4a58-b1ed-ede95f9449bd.png">

It's good practice to keep branch names related to what they are implementing. Of course this matters more as the project & team grows.

jatanolu committed:
Fixed naming and rendering logic

jatanolu committed:
Fixed typo

jatanolu commented:
I am till working, but I have fixed most if not all the names and I changed the logic of the task rendering. @jay-meister 's solution was more of what I was intending when I was originally working through it, so changed it in that way. I will try to have tests and anything else finished up later today.

jay-meister commented on frontend/components/Input.tsx:
```diff
@@ -3,10 +3,10 @@ import { FormEvent, useState } from "react";
-  submited: boolean;
+  submitted: boolean;
   setSubmited: (Submit: boolean) => void;
-  list: ITask[];
-  setList: (list: ITask[]) => void;
+  taskList: ITask[];
+  setTaskList: (taskList: ITask[]) => void;
```
👌

jay-meister commented on frontend/components/RenderTask.tsx:
```diff
@@ -0,0 +1,19 @@
+  id: number;
+  title: string;
+  completed: boolean;
+  func: Function;
```
- as a general rule, try not to use abbreviations when naming, it saves few characters and just introduces additional work for the reader of the code to understand
- more importantly regarding the name `func` or even `function` - in line with previous comments, name this after what it does, not what it is.. A conventional name would be: 'changeHandler'

jay-meister commented on frontend/components/Task.tsx:
```diff
@@ -1,34 +1,44 @@
   return (
     <>
-      <Status {...props} />
+      <h2>Pending</h2>
+      {props.taskList
+        .filter((task) => !task.completed)
+        .map((task) => (
+          <RenderTask {...task} func={handleChange} />
+        ))}
+      <h2>Complete</h2>
+      {props.taskList
+        .filter((task) => task.completed)
+        .map((task) => (
+          <RenderTask {...task} func={handleChange} />
+        ))}
```
Much better 👍

jay-meister commented on frontend/components/Task.tsx:
```diff
@@ -1,34 +1,44 @@
+      {props.taskList
+        .filter((task) => !task.completed)
+        .map((task) => (
+          <RenderTask {...task} func={handleChange} />
```
`func={handleChange}`

This should be a naming red flag to you - if you change names between files, then it becomes harder to keep track.

As mentioned previously, should be: `handleChange={handleChange}`

jay-meister commented on frontend/components/Task.tsx:
```diff
@@ -1,34 +1,44 @@
-    console.log(allTasks);
-    props.setList(allTasks.data.tasks);
+function Task(props: Iprops) {
+  const tasktaskList = async () => {
```
The casing is inconsistent, should be `taskTaskList`.

More importantly, this name is unclear. I have no idea what it does from it's name. Think about what is it doing, then think of a name which captures that.
